src/native/containers/dn-vector-t.h

```diff
@@ -0,0 +1,291 @@
+	DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, default_local_allocator_capacity_size) = 192, \
+	DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, default_local_allocator_byte_size) = ((sizeof (type) * DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, default_local_allocator_capacity_size)) + DN_ALLOCATOR_ALIGN_SIZE (sizeof (dn_vector_t), DN_ALLOCATOR_MEM_ALIGN8) + 32) \
+} DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, sizes); \
+static inline void \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, advance) (DN_DEFINE_VECTOR_IT_T_NAME(name) *it, ptrdiff_t n) \
+{ \
+	dn_vector_it_advance ((dn_vector_it_t *)it, n); \
+} \
+static inline DN_DEFINE_VECTOR_IT_T_NAME(name) \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, next) (DN_DEFINE_VECTOR_IT_T_NAME(name) it) \
+{ \
+	DN_ASSERT ((it.it + 1) < UINT32_MAX); \
+	it.it++; \
+	return it; \
+} \
+static inline DN_DEFINE_VECTOR_IT_T_NAME(name) \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, prev) (DN_DEFINE_VECTOR_IT_T_NAME(name) it) \
+{ \
+	DN_ASSERT (it.it > 0); \
+	it.it--; \
+	return it; \
+} \
+static inline DN_DEFINE_VECTOR_IT_T_NAME(name) \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, next_n) (DN_DEFINE_VECTOR_IT_T_NAME(name) it, uint32_t n) \
+{ \
+	DN_ASSERT ((it.it + n) < UINT32_MAX); \
+	it.it += n; \
+	return it; \
+} \
+static inline DN_DEFINE_VECTOR_IT_T_NAME(name) \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, prev_n) (DN_DEFINE_VECTOR_IT_T_NAME(name) it, uint32_t n) \
+{ \
+	DN_ASSERT (it.it >= n); \
+	it.it -= n; \
+	return it; \
+} \
+static inline type * \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, data) (DN_DEFINE_VECTOR_IT_T_NAME(name) it, uint32_t n) \
+{ \
+	DN_ASSERT (it._internal._vector && it._internal._vector->data); \
+	return (type *)(it._internal._vector->data + n);\
+} \
+static inline DN_DEFINE_VECTOR_T_NAME(name) * \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, custom_alloc) (const DN_DEFINE_VECTOR_CUSTOM_ALLOC_PARAMS_T_SYMBOL_NAME(name) *params) \
+{ \
+	return (DN_DEFINE_VECTOR_T_NAME(name) *)_dn_vector_alloc_capacity (params->allocator, sizeof (type), params->capacity, params->attributes); \
+} \
+static inline DN_DEFINE_VECTOR_T_NAME(name) * \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, alloc) (void) \
+{ \
+	return (DN_DEFINE_VECTOR_T_NAME(name) *)_dn_vector_alloc (DN_DEFAULT_ALLOCATOR, sizeof (type), DN_VECTOR_ATTRIBUTE_INIT_MEMORY); \
+} \
+static inline void \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, free) (DN_DEFINE_VECTOR_T_NAME(name) *vector) \
+{ \
+	dn_vector_free ((dn_vector_t *)vector); \
+} \
+static inline void \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, custom_free) (DN_DEFINE_VECTOR_T_NAME(name) *vector, dn_vector_dispose_func_t dispose_func) \
+{ \
+	dn_vector_custom_free ((dn_vector_t *)vector, dispose_func); \
+} \
+static inline bool \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, custom_init) (DN_DEFINE_VECTOR_T_NAME(name) *vector, DN_DEFINE_VECTOR_CUSTOM_ALLOC_PARAMS_T_SYMBOL_NAME (name) *params) \
+{ \
+	return _dn_vector_init_capacity ((dn_vector_t *)vector, params->allocator, sizeof(type), params->capacity, params->attributes); \
+} \
```

Comment:
So we don't actually use `_dn_vector_custom_params_t:element_size`.  I don't think it should be in the params struct - we should pass it as a regular argument to the custom_init and custom_alloc functions.
I don't think it's really a "customization" in the same way as the other arguments.  You really do have to pass `sizeof(something)`.

I'm suggesting:

```c
struct _ ## name ## _custom_params_t { \
	dn_allocator_t *allocator; \
	uint32_t capacity; \
	uint32_t attributes; \
}; \
... \
```

and

```c
static inline bool
dn_vector_custom_init (
	dn_vector_t *vector,
        uint32_t element_size, 
	const dn_vector_custom_init_params_t *params)
{
	return _dn_vector_init_capacity (vector, params->allocator, element_size, params->capacity, params->attributes);
}
```

(And also I don't want `_dn_vector_init_capacity` to exist.  Just have the  most customizable version of the public function be the one that calls down into the implementation)

Reply:
we could move it into the function signature, I tried to eliminate most of the parameters, but make sense to have the type left as a parameter in the interface.